api/v1beta1/cloudstackaffinitygroup_conversion.go

```diff
@@ -30,3 +33,44 @@ func (dst *CloudStackAffinityGroup) ConvertFrom(srcRaw conversion.Hub) error { /
 	src := srcRaw.(*v1beta2.CloudStackAffinityGroup)
 	return Convert_v1beta2_CloudStackAffinityGroup_To_v1beta1_CloudStackAffinityGroup(src, dst, nil)
 }
+
+//nolint:golint,revive,stylecheck
+func Convert_v1beta1_CloudStackAffinityGroup_To_v1beta2_CloudStackAffinityGroup(in *CloudStackAffinityGroup, out *v1beta2.CloudStackAffinityGroup, s conv.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+
+	csCluster := &v1beta2.CloudStackCluster{}
+	clusterName := in.ObjectMeta.Labels["cluster.x-k8s.io/cluster-name"]
+	namespace := in.ObjectMeta.Namespace
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := v1beta2.K8sClient.Get(context.TODO(), key, csCluster); err != nil {
+		return err
+	}
+
+	out.Spec = v1beta2.CloudStackAffinityGroupSpec{
+		Type:              in.Spec.Type,
+		Name:              in.Spec.Name,
+		ID:                in.Spec.ID,
+		FailureDomainName: csCluster.Spec.FailureDomains[0].Name,
```

Comment:
we might want to check for len(failuredomains) before making this change?

also, what happens if this conversion happens before the csCluster is converted?

Reply:
lastly, why are we hardcoding this affinity group to always be associated with the first failuredomain? Couldn't that be wrong, even for our current upgrade case?

Reply:
> why are we hardcoding this affinity group to always be associated with the first failuredomain? Couldn't that be wrong, even for our current upgrade case?

I'm not sure how it could be wrong. The unique aspects of a failure domain are the account/domain, zone, and endpoint secret to reference.

Affinity groups do not consider zones.

The account/domain and endpoint secret willl be the same across failure domains on upgrade.

Reply:
> we might want to check for len(failuredomains) before making this change?

At least one failure domain should be present or the verification webhook would deny it. I personally prefer not to recheck when we have other assumptions about it, but I don't feel strongly either way.

Reply:
> also, what happens if this conversion happens before the csCluster is converted? I see we're using the v1beta2 client - does that mean failuredomains will always be set? My concern is about getting a v1beta1 csCluster here and throwing a panic

As far as I'm aware, when reading the v1beta1 version of the cluster from ETCD, it will be converted first and other than conversion methods specifically targeting a type, we shouldn't get a v1beta1 type from a get call.

This does introduce the idea of a circular dependency on updated types, but I don't think we have one here.

Reply:
> I'm not sure how it could be wrong.
@rejoshed, could you elaborate further?  If a cluster has multiple FDs defined for it, why does it make sense to arbitrarily pick the first one to associate the AG with?

Reply:
In our upgrade scenario there was only ever one of each of the details that matters to an affinity group.

So, when upgrading without modifications (the limitation we agreed was reasonable) it doesn't matter to an affinity group which failure domain is picked.

And in this case we immediately set this custom resource object to be deleted, so it goes away before more failure domains can be introduced. This is because they would be introduced in a secondary cluster modification. Upgrade and the introduction of additional failure domains are a two step process. 1 to upgrade v1beta1 to v1beta2 and another to add or remove failure domains pointing to additional ACS management host endpoints.